### ISSUE TITLE
Add HOL-Light & s2n-bignum infrastructure to nix environment

### ISSUE THREAD

**Issue by hanno-becker** [pq-code-package/mlkem-native#603]
First HOL-Light proofs for the AArch64 assembly used in mlkem-native are being developed by John Harrison @jargh in https://github.com/jargh/s2n-bignum-dev/tree/mlkem.

We should consider exteding the nix development environment for mlkem-native so that it includes setups of HOL-Light and s2n-bignum ready to check those proofs, both locally and eventually in CI.

- [ ] A non-standard nix-shell `hol_light` (or so) is setup which includes a working joint setup of HOL-Light and s2n-bignum
- [ ] The hol light shell is configurable in terms of the s2n-bignum and hol-light commits used (because those are moving targets)
- [ ] `make -j 16 proofs &` in `s2n-bignum/proofs/arm` works in the shell, running the existing s2n-bignum proofs.

**Comment by potsrevennil** [comment]
Hey, I had now managed to pack hol-light into a nix derivation here https://github.com/potsrevennil/mlkem-c-aarch64/tree/hol_light

Now in both the default nix shell, and the new nix shell, hol_light, one can have access to the hol-light prover.

And I've tested it with a cloned s2n-bignum, in the `arm` directory with my m1, by running `make -j8 p256/bignum_montmul_p256.native`. Unfortunately, running all the proofs on my m1 would have taken too much time, and also eaten up too much memory, I haven't run all the proofs yet.

And this lead to my question, which is how are we going to run the proofs in s2n-bignum in mlkem-native ?

- A naive solution would be to add it as a submodule, but doesn't seem to be ideal
- Adding `s2n-bignum` as a nix derivation would be an option, but we might have to tackle with the Makefile in `s2n-bignum`, since we cannot build artifact into a `/nix/store` after a derivation is built

**Comment by hanno-becker** [comment]
First, thanks a lot @potsrevennil for the fast turnaround!

> Adding s2n-bignum as a nix derivation would be an option, but we might have to tackle with the Makefile in s2n-bignum, since we cannot build artifact into a /nix/store after a derivation is built

I was hoping s2n-bignum could already be cloned as part of the nix derivation, and e.g. an environment variable $S2N_BIGNUM or so be setup pointing to its path.

I don't want to build any proofs as part of the derivation, but only establish an environment that can run the proofs.

**Comment by potsrevennil** [comment]
ok, let me clarify things
What I said about building `s2n-bignum` into a nix derivation is exactly what you said, only clone the source code into the derivation and have an environment variable point to it, and of course this derivation will not run any proofs.

In order to run the proof, we first need to run the make in the directory, which is inside the `/nix/store`. However `/nix/store` is read-only, and since the current makefile of `s2n-bignum` generates some artifacts in the current directory, we'd result in permission denied error

In other words, I'm not sure how I can run the proofs with a s2n-bignum nix derivation

**Comment by mkannwischer** [comment]
Thanks Han, I can confirm that the following works
```
$ nix develop
$ git clone https://github.com/awslabs/s2n-bignum
$ cd s2n-bignum/arm
$ make -j8 p256/bignum_montmul_p256.native
```
Nice!

**Comment by hanno-becker** [comment]
> and since the current makefile of s2n-bignum generates some artifacts in the current directory, we'd result in permission denied error, unfortunately

Could you check what the Makefile is doing here? The main step should be to run HOL-Light on some proof file, which in turn pulls in `arm/proofs/base.ML` -- but that should be OK with read-only files, I would have thought.

**Comment by potsrevennil** [comment]
So running proofs depends on `.o` files:
https://github.com/jargh/s2n-bignum-dev/blob/90cb5e35a823efee15cde72f0237af39a9bf7371/arm/Makefile#L441

And will generate `.native` files as in here:
https://github.com/jargh/s2n-bignum-dev/blob/90cb5e35a823efee15cde72f0237af39a9bf7371/tools/build-proof.sh#L2-L19

We could generate `.o` files during building the nix derivation, but what about `.native` files ?
